### PR TITLE
Allow vectors of ids + types for opq_osm_id

### PR DESCRIPTION
**Pull request by jmaspons**
Requires https://github.com/ropensci/osmdata/pull/282

### CODE REVIEW

**Comment by mpadge** [comment]
Thanks @jmaspons, once you've updated tests for #282, i'll merge that and get straight on to this one.

**Comment by mpadge** [comment]
@jmaspons I see what you're trying to do here, and think that this also is a good improvement. It still needs a bit of work though. Overpass queries should have all of the `type` entries grouped together, whereas your current line here:
``` r
id <- paste(sprintf (" %s(id:%s);\n", opq$id$type, opq$id$id), collapse="")
```
Just puts every ID with its own distinct `type`, so submits queries like `(node(id:1); node(id:2); node(id:3))`. It's more efficient, and thus necessary here, to group these together, to give `(node(id:1,2,3))`. One way to do that would be to replace the line you have above (L#809 fof opq.R) with something like:

``` r
if (!is.null (opq$id)) { # opq with opq_osm_id

    id_types <- data.frame (type = opq$id$type, id = opq$id$id)
    id_types <- lapply (split (id_types, f = factor (id_types$type)), function (i) {
        sprintf (" %s(id:%s);\n", i$type [1], paste0 (i$id, collapse = ","))
    })
    id_types <- paste0 (id_types, collapse = "")
        
    res <- paste0 (opq$prefix, id_types, opq$suffix)

}
```

Note that `split` groups things regardless of their original order, so even if you submit `"way", "rel", "way", "rel"`, they'll still just be split into the appropriate two groups.

---

Another minor issue is that the initial checks also need to be improved. You've currently just added `all` to the `length(id) == 1 && grepl(...)` bit, but that's not quite sufficient here. Those checks need to be replaced with checks to ensure that either `length(type) == 1` *or* `length(type) == length(id)`, and then proceed to ensure all `type` values are the 3 allowed.

---

I think those two tweaks should pretty much get us there - let me know when that's done, and i'll aim to merge asap after that. Thanks!

**Comment by jmaspons** [comment]
I think it's ready.

For the initial checks, I removed the `length(id) == 1` as since https://github.com/ropensci/osmdata/pull/282 opq can work with multiple ids, now also for id in type/id format.

Also, instead of testing `length(type) == length(id)`, I test for `length (id) %% length (type) != 0`, a condition for recycling when creating the data.frame in opq_string

File: R/opq.R [hunk 1]
```diff
@@ -794,8 +806,13 @@ opq_string_intern <- function (opq, quiet = TRUE) {
 
     } else if (!is.null (opq$id)) { # opq with opq_osm_id
 
-        id <- paste (opq$id$id, collapse = ",")
-        id <- sprintf (" %s(id:%s);\n", opq$id$type, id)
+        typeId <- data.frame (type=opq$id$type, id=opq$id$id)
```
Comment (mpadge) [inline R/opq.R:809]:
```suggestion
        type_id <- data.frame (type=opq$id$type, id=opq$id$id)
```

File: R/opq.R [hunk 2]
```diff
@@ -794,8 +806,13 @@ opq_string_intern <- function (opq, quiet = TRUE) {
...
+        typeId <- split(typeId, typeId$type)
```
Comment (mpadge) [inline R/opq.R:810]:
```suggestion
        type_id <- split(type_id, type_id$type)
```

File: R/opq.R [hunk 3]
```diff
@@ -794,8 +806,13 @@ opq_string_intern <- function (opq, quiet = TRUE) {
...
+        id <- mapply (function(type, ids){
+          paste0 (" ", type, "(id:", paste (ids, collapse=","), ");\n")
+        }, type=names(typeId), ids=typeId)
```
Comment (mpadge) [inline R/opq.R:813]:
```suggestion
        }, type=names(type_id), ids=type_id)
```

**Comment by jmaspons** [comment]
snake_case done!

**Pull request merged**